src/client/client.ts

```diff
@@ -0,0 +1,284 @@
+import {
+	ButtonInteraction,
+	Client,
+	ClientEvents,
+	ContextMenuInteraction,
+	Intents,
+	Interaction,
+	Message,
+	SelectMenuInteraction,
+} from "discord.js";
+import { pushPayloads } from "..";
+import type { ClientEvent } from "../interface/clientEvent";
+import type { ContextCommand, MessageCommand, SlashCommand } from "./interfaces";
+import { addGlobalPayload, addGuildPayload } from "./payload";
+
+let instanceClient: Client | undefined;
+let instancePrefix = "!";
+let hasLoggedIn = false;
+const messageCommands: { [index: string]: MessageCommand } = {};
+const slashCommands: { [index: string]: SlashCommand } = {};
+const contextMenus: { [index: string]: (interaction: ContextMenuInteraction) => void } = {};
+const buttons: { [index: string]: (interaction: ButtonInteraction) => void } = {};
+const menus: { [index: string]: (interaction: SelectMenuInteraction) => void } = {};
+
+/**
+ *
+ * @param prefix A prefix to assign to the bot. (default: "!")
+ * @param intents A list of intents to enable on the client. (default: [Intents.FLAGS.GUILDS, Intents.FLAGS.GUILD_MESSAGES, Intents.FLAGS.GUILD_MEMBERS])
+ * @returns Client that has been created
+ */
+function createClient(
+	prefix?: string,
+	intents = [Intents.FLAGS.GUILDS, Intents.FLAGS.GUILD_MESSAGES, Intents.FLAGS.GUILD_MEMBERS]
+): Client {
+	if (prefix) instancePrefix = prefix;
+	instanceClient = new Client({ intents });
+
+	return instanceClient;
+}
+
+/**
+ *
+ * @param token Required token to login the client
+ * @returns Client that has been logged in
+ */
+function loginClient(token: string): Client {
+	if (!instanceClient) throw new Error("No client has been created yet.");
+	if (!token) throw new Error("No token provided.");
+
+	instanceClient
+		.login(token)
+		.then(() => (hasLoggedIn = true))
+		.catch(console.error.bind(console));
+
+	registerEvent({
+		name: "messageCreate",
+		once: false,
+		callback: (message: Message) => {
+			const args = message.content.split(" ");
+			const command = args.shift();
+
+			if (command && command.startsWith(instancePrefix)) {
+				const messageCommandData = messageCommands[command.substring(instancePrefix.length)];
+
+				if (messageCommandData != undefined) {
+					const author = message.member;
+					if (!author) return;
+
+					// flag check
+					const flags = messageCommandData.permissions.flags;
+					if (flags && author.permissions.toArray().some(perm => flags.has(perm))) {
+						messageCommandData.callback(message, args);
+						return;
+					}
+
+					// role check
+					const allowed = messageCommandData.permissions.allowed;
+					const denied = messageCommandData.permissions.denied;
+
+					const hasPermission = author.roles.cache.some(role => allowed.has(role.id));
+					const isDenied = denied == undefined ? false : author.roles.cache.some(role => denied.has(role.id));
+
+					if (isDenied) return;
+					if (hasPermission) {
+						messageCommandData.callback(message, args);
+						return;
+					}
```

Comment:
Could make `messageCommandData.permissions` optional to allow global unrestricted `messageCommand`